GeofenceDisplay/src/geofence-display.ts

```diff
@@ -1,155 +1,111 @@
-    // })
-
-    // // GeofencePane related code
-    // // TODO: Tie in with existing setup
-    // _map.createPane('geofence')
-    // let geofencePane = _map.getPane('geofence')
-    // geofencePane.style.zIndex = 230
-    // geofencePane.style.pointerEvents = 'none'
-
-    // _map.createPane('geofenceLabel')
-    // let geofenceLabelPane = _map.getPane('geofenceLabel')
-    // geofenceLabelPane.style.zIndex = 235
-    // geofenceLabelPane.style.pointerEvents = 'none'
-
-    // self.layer.geofence = L.layerGroup(null, { pane: 'geofence' }).addTo(_map)
-    // self.layer.geofence.name$ = 'geofence'
-
-    // //
-
-    // // TODO: show/clear geofence functionality required here?
-    // self.showGeofence = function (geofence, show) {
-    //   if (null == geofence) {
-    //     return
-    //   }
-
-    //   show = !!show
-
-    //   if (true === show) {
-    //     geofence.show(self.layer.geofence)
-    //   } else if (false === show) {
-    //     geofence.hide()
-    //   }
-    // }
-
-    // self.clearGeofences = function () {
-    //   for (let geofenceID in self.geofence.map) {
-    //     let geofence = self.geofence.map[geofenceID]
-    //     delete self.geofence.map[geofenceID]
-    //     if (geofence && geofence.hide) {
-    //       geofence.hide()
-    //     }
-    //   }
-    // }
-
-    // // TODO: add geofencePane to div
-
-    return _map
+    let self = this
+    self.options.debug && console.log('GeofenceDisplay onAdd function called.')
+    self.options.debug &&
+      console.log(
+        'Local geofence variable before add:',
+        self._state.geofenceByID
+      )
+    self.options.geofences.forEach((fence: any) => {
+      let geo = self.options.geofences[fence]
+      let geofence = new Geofence(geo, { map: _map })
+      self._state.geofenceByID[geo.id] = geofence
+      geofence.show()
+    })
+    self.options.debug &&
+      console.log(
+        'Local geofence variable after add:',
+        self._state.geofenceByID
+      )
   },
 
   onRemove: function (this: any, _map: any) {
-    console.log('PGeoD onRemove')
-  },
-
-  clear: function (this: any) {
-    console.log('PGeoD clear')
+    let self = this
+    self.options.debug &&
+      console.log('GeofenceDisplay onRemove function called.')
+    self.options.debug &&
+      console.log(
+        'Local geofence variable before remove:',
+        self._state.geofenceByID
+      )
+    for (let id in self._state.geofenceByID) {
+      self._state.geofenceByID[id].hide()
+      delete self._state.geofenceByID[id]
+    }
+    self.options.debug &&
+      console.log(
+        'Local geofence variable after remove:',
+        self._state.geofenceByID
+      )
   },
 })
 
 class Geofence {
-  ent = null
-  ctx = null
-  poly = L.polygon(
-    [
-      [37, -109.05],
-      [41, -109.03],
-      [41, -102.05],
-      [37, -102.04],
-    ],
-    { color: 'red' }
-  )
+  ent: any = null
+  ctx: any = null
+  poly: any = null
```

Comment:
I wonder would this work:
```
poly: ReturnType<typeof L.polygon>
```
null is should not be needed, because poly should be created in the constructor, that way the poly member var is awlays valid

Reply:
In the pqam code, poly isn't in the constructor at all, and is only created if `null == this.poly` once the show() method has been called - are we changing this to create the poly in the constructor?

In any case, the property `poly: ReturnType<typeof L.polygon>` with a constructor featuring the line `this.poly = L.polygon(poly {color: this.ent.colour})` doesn't seem to pose any problems.

Reply:
Answer given by this comment: https://github.com/plantquest/leaflet-plugins/pull/8/files#r1392385707